### PR TITLE
feat: split validation statistics by task name

### PR DESCRIPTION
**Pull request by yuki-97**
1. Support split validation statistics by task name in SFT/GRPO/Distillation by using multiple val dataloaders like RM/DPO.
2. Support set custom task name, it's useful for saving checkpoint according to a specific dataset's val statistic (e.g. accuracy, val_loss).

image

## Summary by CodeRabbit

* **New Features**
  * Multi-task validation support with per-dataset accuracy and loss tracking for checkpointing
  * Per-dataset task naming for validation metrics (e.g., `val:accuracy_<TaskName>`)

* **Documentation**
  * Added guidance for configuring dataset-specific validation metrics in training configs
  * Included YAML examples for multi-dataset validation setups

* **Tests**
  * Updated validation metric checks to include per-task accuracy metrics

### CODE REVIEW

File: nemo_rl/algorithms/sft.py [hunk 1]
```diff
@@ -245,100 +261,119 @@ def validate(
...
+            total_val_loss += task_val_loss
+            total_num_valid_tokens += task_num_valid_tokens
+
+            if val_batches > 0 and cur_batch_idx >= val_batches - 1:
                 break
 
-        if sum_num_valid_tokens > 0:
-            val_metrics["val_loss"] /= sum_num_valid_tokens
+        # Calculate validation metrics
+        if total_num_valid_tokens > 0:
```
Comment (terrykong) [inline nemo_rl/algorithms/sft.py:348]: nit:

```suggestion
        if total_num_valid_tokens > 0:
            assert "total" not in val_loss, f"total is a reserved task_name since it is used in the metrics as the aggregate label"
```

File: nemo_rl/algorithms/sft.py [hunk 2]
```diff
@@ -511,33 +546,27 @@ def sft_train(
...
                 ):
+                    # +1 because step is 0-indexed
                     sft_save_state["step"] = (current_step + 1) % len(train_dataloader)
                     sft_save_state["total_steps"] = total_steps + 1
                     sft_save_state["epoch"] = current_epoch
                     sft_save_state["total_valid_tokens"] = total_valid_tokens
 
                     full_metric_name = master_config["checkpointing"]["metric_name"]
                     if full_metric_name is not None:
-                        assert full_metric_name.startswith(
```
Comment (terrykong) [inline nemo_rl/algorithms/sft.py:531]: why remove this assert?

File: nemo_rl/data/datasets/raw_dataset.py [hunk 3]
```diff
@@ -14,19 +14,48 @@
...
+
     dataset: Dataset
     # `val_dataset` is used only when current dataset is used for both training and validation
     val_dataset: Dataset | None
-    processor: TaskDataProcessFnCallable
+    task_name: str
     task_spec: TaskDataSpec
+    processor: TaskDataProcessFnCallable
+
+    def common_init(
```
Comment (terrykong) [inline nemo_rl/data/datasets/raw_dataset.py:36]: any reason not to do the more familiar `__init__` so that subclasses can call super().__init__(...)?

File: nemo_rl/algorithms/sft.py (hunk 1 above)
Comment (yuki-97) [reply nemo_rl/algorithms/sft.py:348]: https://github.com/NVIDIA-NeMo/RL/commit/ef3e6795ae1bcbf8e1c0f10e28f9e47821acee1d

File: nemo_rl/algorithms/sft.py (hunk 2 above)
Comment (yuki-97) [reply nemo_rl/algorithms/sft.py:531]: not removed, just moved to `nemo_rl/utils/checkpoint.py`. :)
so that we could assert this at the beginning and reduce duplicated things in different algos.

File: nemo_rl/data/datasets/raw_dataset.py (hunk 3 above)
Comment (yuki-97) [reply nemo_rl/data/datasets/raw_dataset.py:36]: as discussed offline, keep it as is.

the param of `__init__` will be different between subclasses and base class.
